File: 08/1.py
def load_data():
    with open('data.txt', 'r') as f:
        data = f.read().splitlines()
    return data


def parse_instruction(line):
    return line[:3], int(line[4:])


def step_seen_before(history, i):
    return i in history


def get_pos_and_acc_movements(line):
    code, value = parse_instruction(line)
    if code == 'acc':
        movements = (1, value)
    if code == 'nop':
        movements = (1, 0)
    if code == 'jmp':
        movements = (value, 0)
    return movements


def increment_pos_and_acc_using_instruction(instruction, pos, acc):
    pos_delta, acc_delta = get_pos_and_acc_movements(instruction)
    pos += pos_delta
    acc += acc_delta
    return pos, acc


def main():
    data = load_data()

    acc, pos, history = 0, 0, []
    while True:
        if step_seen_before(history, pos):
            break
        
        history.append(pos)
        instruction = data[pos]
        pos, acc = increment_pos_and_acc_using_instruction(instruction, pos ,acc)
    
    print(acc)
    return


if __name__ == '__main__':
    main()

File: 08/1.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

func loadFile(filename string) []string {
	data, err := ioutil.ReadFile(filename)
	if err != nil {
		panic(err)
	}
	strValues := strings.Split(strings.Replace(string(data), "\r", "", -1), "\n")
	return strValues
}

func parseInstruction(line string) (string, int) {
	code := line[:3]
	value, _ := strconv.Atoi(line[4:])
	return code, value
}

func getPosAndAccMovements(line string) (int, int) {
	code, value := parseInstruction(line)
	pos_delta, acc_delta := 1, 0
	switch code {
	case "acc":
		acc_delta = value
	case "jmp":
		pos_delta = value
	}
	return pos_delta, acc_delta
}

func contains(slice []int, element int) bool {
	for _, a := range slice {
		if a == element {
			return true
		}
	}
	return false
}

func main() {
	strValues := loadFile("data.txt")
	var history []int
	acc, pos := 0, 0
	for {
		if contains(history, pos) {
			break
		}

		history = append(history, pos)
		line := strValues[pos]
		pos_delta, acc_delta := getPosAndAccMovements(line)
		pos += pos_delta
		acc += acc_delta
	}
	fmt.Println(acc)
}
